src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY", None)
+        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY", None)
+        no_proxy = os.environ.get("JUJU_CHARM_NO_PROXY", None)
+
+        self.proxies: ProxySetting = {}
+        if http_proxy:
+            self.proxies["http"] = http_proxy
+        if https_proxy:
+            self.proxies["https"] = https_proxy
+        if no_proxy:
+            self.proxies["no_proxy"] = no_proxy
+
+        self.session = requests.Session()
+        # TODO: Review the retry strategy
+        adapter = requests.adapters.HTTPAdapter(
+            max_retries=urllib3.Retry(
+                total=10, backoff_factor=0.3, status_forcelist=[500, 502, 503, 504]
+            )
+        )
+        self.session.mount("http://", adapter)
+        self.session.mount("https://", adapter)
+        if self.proxies:
+            # setup proxy for requests
+            self.session.proxies.update(self.proxies)
+            # add proxy to fastcore which ghapi uses
+            proxy = urllib.request.ProxyHandler(self.proxies)
+            opener = urllib.request.build_opener(proxy)
+            fastcore.net._opener = opener
+
+        self._github = GhApi(token=token)
+        self._jinja = jinja2.Environment(
+            loader=jinja2.FileSystemLoader("templates"), autoescape=True
+        )
+        self._lxd = pylxd.Client()
+
+        if "/" in path:
+            owner, repo = path.split("/")
+            self.path = GitHubRepo(owner=owner, repo=repo)
+        else:
+            self.path = GitHubOrg(org=path)
+
+        self.app_name = app_name
+        self.reconcile_interval = reconcile_interval
+        self.image = image
+
+    @classmethod
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def install_deps(cls) -> None:
+        """Install dependencies."""
+
+        logger.info("Installing charm dependencies.")
+
+        # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
+        execute_command(["/usr/bin/apt", "remove", "-qy", "lxd", "lxd-client"], check=False)
+        execute_command(["/usr/bin/snap", "install", "lxd", "--channel=latest/stable"])
+        execute_command(["/usr/bin/snap", "refresh", "lxd", "--channel=latest/stable"])
+        execute_command(["/snap/bin/lxd", "waitready"])
+        execute_command(["/snap/bin/lxd", "init", "--auto"])
+        execute_command(["/usr/bin/chmod", "a+wr", "/var/snap/lxd/common/lxd/unix.socket"])
+        execute_command(["/snap/bin/lxc", "network", "set", "lxdbr0", "ipv6.address", "none"])
+        execute_command(
+            [
+                "/usr/bin/apt",
+                "install",
+                "-qy",
+                "cpu-checker",
+                "libvirt-clients",
+                "libvirt-daemon-driver-qemu",
+            ],
+        )
+        cls.runner_bin_path.parent.mkdir(parents=True, exist_ok=True)
+
+        logger.info("Finished installing charm dependencies.")
```

Comment:
Are we guaranteed to have successfully installed all of these dependencies?

Reply:
Failure to install should result in an exception due to the command returning non-zero exit code.

I have move this to charm.py and will handle the exception there.